python/sglang/srt/managers/controller/infer_batch.py

```diff
@@ -947,28 +947,56 @@ def init_flashinfer_args(
 ):
     """Init auxiliary variables for FlashInfer attention backend."""
     num_qo_heads = model_runner.model_config.num_attention_heads // model_runner.tp_size
-    num_kv_heads = model_runner.model_config.get_num_kv_heads(model_runner.tp_size)
+    # NOTE (yifan): we partitioned K and V along both TP and SP dimensions.
+    # And here tp_size represents actual TP size * SP size.
+    num_kv_heads = model_runner.model_config.get_num_kv_heads(
+        model_runner.tp_size // model_runner.sp_size
+    )
     head_dim = model_runner.model_config.head_dim
     batch_size = len(req_pool_indices)
 
     if forward_mode == ForwardMode.DECODE:
         paged_kernel_lens = seq_lens
     else:
+        # FIXME (yifan): Simplify the layout adjust code below.
+        seq_lens_cpu = seq_lens.cpu().numpy()
+        prefix_lens_cpu = prefix_lens.cpu().numpy()
+        for i in range(batch_size):
+            seq_len = seq_lens_cpu[i]
+            prefix_len = prefix_lens_cpu[i]
+            seq_lens_cpu[i] = seq_len // model_runner.sp_size + (
+                seq_len % model_runner.sp_size > model_runner.sp_rank
+            )
+            prefix_lens_cpu[i] = prefix_len // model_runner.sp_size + (
+                prefix_len % model_runner.sp_size > model_runner.sp_rank
+            )
```

Comment:
now we don't use this layout.
Instead, for all `sp_rank != sp_size - 1`, the local len is `ceil(len / sp_size)`; for `sp_rank == sp_size - 1`, it's the `len % ceil(len / sp_size)`

Reply:
Got it. Let me merge #3 and get this fixed

Reply:
Thanks! I have a fix for this locally and will push it soon.